p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,516 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/connmgr"
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/pnet"
+	"github.com/libp2p/go-libp2p/core/sec"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	"github.com/libp2p/go-libp2p/p2p/security/noise"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal/encoding"
+
+	logging "github.com/ipfs/go-log/v2"
+	ma "github.com/multiformats/go-multiaddr"
+	mafmt "github.com/multiformats/go-multiaddr-fmt"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multihash"
+	pionlogger "github.com/pion/logging"
+
+	"github.com/pion/webrtc/v3"
+)
+
+var log = logging.Logger("webrtc-transport")
+
+var dialMatcher = mafmt.And(mafmt.UDP, mafmt.Base(ma.P_WEBRTC), mafmt.Base(ma.P_CERTHASH))
+
+const (
+	// handshakeChannelNegotiated is used to specify that the
+	// handshake data channel does not need negotiation via DCEP.
+	// A constant is used since the `DataChannelInit` struct takes
+	// references instead of values.
+	handshakeChannelNegotiated = true
+	// handshakeChannelId is the agreed ID for the handshake data
+	// channel. A constant is used since the `DataChannelInit` struct takes
+	// references instead of values. We specify the type here as this
+	// value is only ever copied and passed by reference
+	handshakeChannelId = uint16(0)
+)
+
+// timeout values for the peerconnection
+// https://github.com/pion/webrtc/blob/v3.1.50/settingengine.go#L102-L109
+const (
+	DefaultDisconnectedTimeout = 5 * time.Second
+	DefaultFailedTimeout       = 25 * time.Second
+	DefaultKeepaliveTimeout    = 2 * time.Second
+)
+
+type WebRTCTransport struct {
+	webrtcConfig webrtc.Configuration
+	rcmgr        network.ResourceManager
+	privKey      ic.PrivKey
+	noiseTpt     *noise.Transport
+	localPeerId  peer.ID
+
+	// timeouts
+	peerConnectionTimeouts IceTimeouts
+
+	// in-flight connections
+	maxInFlightConnections uint32
+}
+
+var _ tpt.Transport = &WebRTCTransport{}
+
+type Option func(*WebRTCTransport) error
+
+type IceTimeouts struct {
+	Disconnect time.Duration
+	Failed     time.Duration
+	Keepalive  time.Duration
+}
+
+// WithPeerConnectionIceTimeouts sets the ice disconnect, failure and keepalive timeouts
+func WithPeerConnectionIceTimeouts(timeouts IceTimeouts) Option {
```

Comment:
This would probably be nicer as 3 separate options, but I don't really understand what they're used for in the first place, so I can't really tell.

Reply:
Will take some time tomorrow to either revisit this myself, or explain why my initial reasoning still stand.

Reply:
@ckousik mind helping me elaborate here? I am honestly not too sure how to explain this.
Either way I do not mind splitting up, but perhaps first we need to make sure we do really want/need it.